File: LeetCode/20_Valid_Parentheses.js
var isValid = function(s) {
    var stack = [];
    const table = { '(':')', '{':'}', '[':']' };

    for (let i = 0; i < s.length; i++) {
        let par = s[i];
        if (table[par]) {
            stack.push(par);
        }
        else {
            if (stack.length == 0) {
                return false;
            }
            
            let z = stack.pop();
            if (par !== table[z]) {
                return false;
            } 
        }
    } console.log(stack.length == 0);
};

s = "()[]{}"
// s = "()"
isValid(s)

File: LeetCode/20_Valid_Parentheses.py
'''
Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.

Input: s = "()"
Output: true

Input: s = "()[]{}"
Output: true

Input: s = "(]"
Output: false
'''
def isValid(s):
    stack = []
    
    table = { '(':')', '{':'}', '[':']' }
    for par in s:
        if par in table.keys():
            stack.append(par)
        else:
            if len(stack) == 0:
                return False
            z = stack.pop()
            if par != table[z]:
                return False
    return len(stack) == 0


s = "()[]{}"
print(isValid(s))